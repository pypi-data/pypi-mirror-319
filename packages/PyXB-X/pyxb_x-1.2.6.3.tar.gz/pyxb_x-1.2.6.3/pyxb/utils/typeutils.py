import datetime

def make_base_dt(value):
    """
    Create a copy of `value` using datetime baseclass. This allows us to do datetime arithmetic on Python 3.8+
    Args:
        value (pyxb.binding.datatypes.datetime): PyXB datetime object

    Returns:
        datetime.datetime: Matching datetime base instance
    """
    return datetime.datetime(
        value.year, 
        value.month, 
        value.day, 
        value.hour, 
        value.minute, 
        value.second, 
        value.microsecond, 
        tzinfo=value.tzinfo
    )
