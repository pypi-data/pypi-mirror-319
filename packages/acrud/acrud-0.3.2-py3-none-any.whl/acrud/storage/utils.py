import os


def get_meta_data_file_path(file_path: str):
    file_path = file_path.split(".")[:-1]
    file_path = ".".join(file_path)
    meta_data_file_path = file_path + "_meta.json"
    return meta_data_file_path
