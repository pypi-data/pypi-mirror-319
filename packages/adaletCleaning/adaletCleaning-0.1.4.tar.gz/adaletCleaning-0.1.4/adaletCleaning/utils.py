import re
import xml.etree.ElementTree as elemTree
import os

def consonantConsecutiveList(word, count):
    consonant_list = re.split(r"[aeıioöuü]+", word, flags=re.I)
    return [y for y in consonant_list if len(y) > count]


def vowelConsecutiveList(word, count):
    consonant_list = re.split(r"[bcdfghjklmnprstvyz]+", word, flags=re.I)
    return [y for y in consonant_list if len(y) > count]


def turkish_lower(text):
    text = text.replace("T.C.", " tc ")
    text = re.sub(r'İ', 'i', text)
    text = re.sub(r'I', 'ı', text)
    text = text.lower()
    return text


def turkish_upper(text):
    text = re.sub(r'i', 'İ', text)
    text = text.upper()
    return text



#: read a xml file and parse its roots
def getCityDistrictNames():
    current_dir = os.path.dirname(__file__)  # Geçerli dosyanın bulunduğu dizin

    il_ilceler_path = os.path.join(current_dir, "resources", "il_ilceler.xml")

    tree = elemTree.parse(il_ilceler_path)
    root = tree.getroot()

    cityNames = []
    districtNames = []
    cityDistrictNames = []

    try:
        for city in root.findall('CITY'):
            cityName = turkish_lower(str(city.attrib['cityname']))
            cityNames.append(cityName)
            for district in city.findall('DISTRICT'):
                districtName = turkish_lower(district.find('DISTNAME').text)
                if len(districtName.split('/')) > 1:
                    districtName = districtName.split('/')[0].strip()
                if len(districtName.split(' ')) > 1:
                    districtName = districtName.split(' ')[1].strip()
                districtNames.append(districtName)

                cityDistrictNames.append(cityName)
                cityDistrictNames.append(districtName)

    except Exception as e:
        print(e)

    return list(set(cityNames)), list(set(districtNames)), list(set(cityDistrictNames))

