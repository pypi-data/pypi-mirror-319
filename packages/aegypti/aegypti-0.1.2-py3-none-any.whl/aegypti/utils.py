# Created on 01/09/2025
# Author: Frank Vega

import scipy.sparse as sparse
import networkx as nx
import numpy as np
import random
import string


def generate_short_hash(length=6):
    """Generates a short random alphanumeric hash string.

    Args:
        length: The desired length of the hash string (default is 6).

    Returns:
        A random alphanumeric string of the specified length.
        Returns None if length is invalid.
    """

    if not isinstance(length, int) or length <= 0:
        print("Error: Length must be a positive integer.")
        return None

    characters = string.ascii_letters + string.digits  # alphanumeric chars
    return ''.join(random.choice(characters) for i in range(length))

def make_symmetric(matrix):
    """Makes an arbitrary sparse matrix symmetric efficiently.

    Args:
        matrix: A SciPy sparse matrix (e.g., csc_matrix, csr_matrix, etc.).

    Returns:
        scipy.sparse.csc_matrix: A symmetric sparse matrix.
    Raises:
        TypeError: if the input is not a sparse matrix.
    """

    if not sparse.issparse(matrix):
        raise TypeError("Input must be a SciPy sparse matrix.")

    rows, cols = matrix.shape
    if rows != cols:
        raise ValueError("Matrix must be square to be made symmetric.")

    # Convert to COO for efficient duplicate handling
    coo = matrix.tocoo()

    # Concatenate row and column indices, and data with their transposes
    row_sym = np.concatenate([coo.row, coo.col])
    col_sym = np.concatenate([coo.col, coo.row])
    data_sym = np.concatenate([coo.data, coo.data])

    # Create the symmetric matrix in CSC format
    symmetric_matrix = sparse.csc_matrix((data_sym, (row_sym, col_sym)), shape=(rows, cols))
    symmetric_matrix.sum_duplicates() #sum the duplicates

    return symmetric_matrix

def random_matrix_tests(matrix_shape, sparsity=0.9):
    """
    Performs random tests on a sparse matrix.

    Args:
        matrix_shape (tuple): Shape of the matrix (rows, columns).
        num_tests (int): Number of random tests to perform.
        sparsity (float): Sparsity of the matrix (0.0 for dense, close to 1.0 for very sparse).

    Returns:
        list: A list containing the results of each test.
        sparse matrix: the sparse matrix that was tested.
    """

    rows, cols = matrix_shape
    size = rows * cols

    # Generate a sparse matrix using random indices and data
    num_elements = int(size * (1 - sparsity))  # Number of non-zero elements
    row_indices = np.random.randint(0, rows, size=num_elements)
    col_indices = np.random.randint(0, cols, size=num_elements)
    data = np.ones(num_elements, dtype=np.int8)

    sparse_matrix = sparse.csc_matrix((data, (row_indices, col_indices)), shape=(rows, cols))

    symmetric_matrix = make_symmetric(sparse_matrix)  

    symmetric_matrix.setdiag(0)

    return symmetric_matrix

def generate_triangles_from_edges(adjacency_matrix, triangles):
    """
    Optimized version: Generate triangles given a list of edge pairs.
    Avoids redundant set creation.

    Args:
        adjacency_matrix: A SciPy sparse adjacency matrix.
        triangles: A list of tuples, where each tuple (u, v) represents an edge.

    Returns:
        All triangles formed using at least on side in the given edges.
        Raises TypeError if inputs are not of the correct type.
        Raises ValueError if the input matrix is not square or vertex indices are out of range.
    """
    if not sparse.isspmatrix(adjacency_matrix):
        raise TypeError("adjacency_matrix must be a SciPy sparse matrix.")
    if not all(isinstance(edge, tuple) and len(edge) == 2 for edge in triangles):
        raise TypeError("Each element in triangles must be a 2-tuple.")

    rows, cols = adjacency_matrix.shape
    if rows != cols:
        raise ValueError("Input matrix must be square.")

    visited = set()
    for current_node, neighbor in triangles:
        if not (0 <= current_node < adjacency_matrix.shape[0] and 0 <= neighbor < adjacency_matrix.shape[0]):
            raise ValueError("Vertex indices in triangles are out of range.")
        current_row_indices = adjacency_matrix.getrow(current_node).indices
        neighbor_row_indices = adjacency_matrix.getrow(neighbor).indices

        i = j = 0
        while i < len(current_row_indices) and j < len(neighbor_row_indices):
            if current_row_indices[i] == neighbor_row_indices[j]:
                minimum = min(current_node, neighbor, current_row_indices[i])
                maximum = max(current_node, neighbor, current_row_indices[i])
                betweenness = set([current_node, neighbor, current_row_indices[i]]) - {minimum, maximum}
                if betweenness:
                  new_triangle = (str(minimum), str(next(iter(betweenness))), str(maximum))
                  if new_triangle not in visited:
                    visited.add(new_triangle)
                i += 1
                j += 1
            elif current_row_indices[i] < neighbor_row_indices[j]:
                i += 1
            else:
                j += 1
    return visited

def edges_to_sparse_matrix(edges, num_vertices=None, is_directed=False):
    """
    Converts a set of edges to a SciPy sparse matrix.

    Args:
        edges: A list or set of tuples/lists representing edges.
        num_vertices: The total number of vertices (optional, inferred if None).
        is_directed: Whether the graph is directed (default: False).

    Returns:
        A SciPy CSR sparse matrix.
    """

    if not edges: #Handle empty edge case
        if num_vertices is None:
            raise ValueError("Number of vertices must be specified if edges are empty")
        return sparse.csr_matrix((num_vertices, num_vertices))

    rows = []
    cols = []
    data = []

    for u, v in edges:
        rows.append(u)
        cols.append(v)
        data.append(1)  # Assuming unweighted graph

        if not is_directed:
            rows.append(v)
            cols.append(u)
            data.append(1)

    if num_vertices is None:
        num_vertices = max(max(row, col) for row, col in zip(rows, cols)) + 1
    
    return sparse.csr_matrix((data, (rows, cols)), shape=(num_vertices, num_vertices))

def sparse_matrix_to_edges(adj_matrix, is_directed=False):
    """
    Converts a SciPy sparse adjacency matrix to a set of edges.

    Args:
        adj_matrix: A SciPy sparse adjacency matrix.
        is_directed: Whether the matrix represents a directed graph (default: False).

    Returns:
        A set of tuples representing the edges.
    """

    edges = set()
    rows, cols = adj_matrix.nonzero()

    for i, j in zip(rows, cols):
        edges.add((i, j))
        if not is_directed:
            edges.add((j, i))

    return edges

def sparse_matrix_to_edges_efficient(adj_matrix, is_directed=False):
    """
    Converts a SciPy sparse adjacency matrix to a set of edges efficiently.

    Args:
        adj_matrix: A SciPy sparse adjacency matrix.
        is_directed: Whether the matrix represents a directed graph (default: False).

    Returns:
        A set of tuples representing the edges.
    """

    edges = set()
    rows, cols = adj_matrix.nonzero()
    if is_directed:
        for i, j in zip(rows, cols):
            edges.add((i, j))
    else:
        for i, j in zip(rows, cols):
            if i <= j: # Avoid duplicates in undirected graphs
                edges.add((i, j))
    return edges

def nx_to_sparse(graph, format='csr'):
    """
    Converts a NetworkX graph to a SciPy sparse adjacency matrix.

    Args:
        graph: The NetworkX graph.
        format: The desired sparse matrix format (e.g., 'csr', 'csc', 'coo', 'lil', 'dok'). Default is 'csr'.

    Returns:
        A SciPy sparse adjacency matrix in the specified format, or None if the graph is empty.
    Raises:
        ValueError: If the specified format is invalid.
    """

    if not graph:
        return None

    adj_matrix = nx.to_scipy_sparse_array(graph, format=format)
    return adj_matrix


def nx_to_sparse_alternative(graph):
    """
    Converts a NetworkX graph to a SciPy sparse adjacency matrix using an alternative method.

    Args:
        graph: The NetworkX graph.

    Returns:
        A SciPy CSR sparse adjacency matrix, or None if the graph is empty.
    """
    if not graph:
        return None
    adj_matrix = nx.adjacency_matrix(graph)
    return adj_matrix

def find_triangle_coordinates_brute_force(adjacency_matrix):
    """
    Finds the coordinates of all triangles in a given SciPy sparse matrix.

    Args:
        adjacency_matrix: A SciPy sparse matrix (e.g., csr_matrix).
    
    Returns:
        A list of tuples, where each tuple represents the coordinates of a triangle.
        A triangle is defined by three non-zero entries forming a closed loop.
    """

    if not sparse.isspmatrix(adjacency_matrix):
        raise TypeError("Input must be a SciPy sparse matrix.")
    
    rows, cols = adjacency_matrix.shape
    if rows != cols:
        raise ValueError("Input matrix must be square.")
    
    n = adjacency_matrix.shape[0]
    visited = set()
    for i in range(n-2):
        for j in range(i + 1, n-1):
            if adjacency_matrix[i, j]:  # Check if edge (i, j) exists
                for k in range(j + 1, n):
                    if adjacency_matrix[i, k] and adjacency_matrix[j, k]:  # Check if edges (i, k) and (j, k) exist
                         visited.add((str(i), str(j), str(k)))
    return visited

def string_simple_format(is_free):
  """
  Returns a string indicating whether a graph is triangle-free.

  Args:
    is_free: An Boolean value, True if the graph is triangle-free, False otherwise.

  Returns:
    "Triangle Free" if triangle is True, "Triangle Found" otherwise.
  """
  return "Triangle Free" if is_free  else "Triangle Found"

def string_complex_format(triangle):
  """
  Returns a string indicating all the triangles found in a graph.

  Args:
    triangle: A tuple value, 
    None if the graph is triangle-free,
    a tuples of triangle vertices otherwise.
  
  Returns:
    "Triangle Free" if triangle is None, "Triangles Found (a, b, c)" otherwise.
  """
  if triangle:
      if isinstance(triangle, int):
        return f"Cover Size {triangle}"
      return f"Triangle Found {triangle}"
  else:
     return "Triangle Free"