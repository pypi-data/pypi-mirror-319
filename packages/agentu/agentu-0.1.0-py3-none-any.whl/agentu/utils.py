from typing import Dict, Any
import json

def validate_json_response(response: str) -> Dict[str, Any]:
    """Validate and parse JSON response from Ollama."""
    try:
        data = json.loads(response)
        required_keys = {"selected_tool", "parameters", "reasoning"}
        if not all(key in data for key in required_keys):
            raise ValueError("Missing required keys in response")
        return data
    except json.JSONDecodeError as e:
        raise ValueError(f"Invalid JSON response: {str(e)}")
