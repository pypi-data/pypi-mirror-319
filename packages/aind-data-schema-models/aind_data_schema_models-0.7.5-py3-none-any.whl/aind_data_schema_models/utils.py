""" General utilities for constructing models from CSV files """

import re
from pydantic import BaseModel, Field
from typing import Union, List, Type, Any
from typing_extensions import Annotated


def to_class_name_underscored(name: str) -> str:
    """Convert a name to a valid class name by capitalizing and removing non-alphanumeric characters."""
    return "_" + re.sub(r"\W+", "_", name.title()).replace(" ", "")


def to_class_name(name: str) -> str:
    """Convert a name to a valid class name by capitalizing and removing non-alphanumeric characters."""
    return re.sub(r"\W|^(?=\d)", "_", name.title()).replace(" ", "")


def one_of_instance(instances: List[Type[BaseModel]], discriminator="name") -> Annotated[Union[Any], Field]:
    """
    Make an annotated union of class instances
    Parameters
    ----------
    instances : List[Type[BaseModel]]
      A list of class instances.
    discriminator : str
      Each model in instances should have a common field name where each item
      is unique to the model. This will allow pydantic to know which class
      should be deserialized. Default is 'name'.

    Returns
    -------
    Annotated[Union[Any], Field]
      An annotated field that can be used to define a type where a choice from a
      possible set of classes can be selected.

    """
    return Annotated[Union[tuple(type(i) for i in instances)], Field(discriminator=discriminator)]
