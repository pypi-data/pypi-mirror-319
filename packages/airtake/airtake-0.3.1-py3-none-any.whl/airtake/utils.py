from time import time_ns
from random import random
from uuid import uuid4

alphabet = '_-0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'

def generate_id() -> str:
  alphabet_len = len(alphabet)

  id = ''
  for _ in range(32):
      id += alphabet[int(random() * alphabet_len) | 0]
  return id

def generate_device_id() -> str:
  return f'$device:{uuid4()}'

def timestamp() -> int:
  return time_ns() // 1_000_000
