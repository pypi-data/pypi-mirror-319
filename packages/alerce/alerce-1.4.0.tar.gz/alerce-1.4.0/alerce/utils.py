from pandas import DataFrame, read_csv
from io import StringIO
from astropy.table import Table
from .exceptions import handle_error, FormatValidationError
import requests
import abc


class Result(abc.ABC):
    def __init__(self, format="json"):
        self.format = format

    @abc.abstractmethod
    def to_pandas(self, index=None, sort=None):
        """Convert the result to a pandas dataframe

        :index: index for the pandas dataframe
        :sort: sorting column for the dataframe
        :returns: the processed dataframe

        """
        pass

    @abc.abstractmethod
    def to_votable(self):
        pass

    @abc.abstractmethod
    def to_json(self):
        pass

    @abc.abstractmethod
    def to_csv(self):
        pass

    def result(self, index=None, sort=None):
        """Creates the result depending on the arguments and the expected format

        :index: if format is pandas, this is the index column
        :sort: if format is pandas, this is the sort column
        :returns: Result in the indicated format

        """
        if self.format == "pandas":
            return self.to_pandas(index, sort)
        elif self.format == "votable":
            return self.to_votable()
        elif self.format == "csv":
            return self.to_csv()
        elif self.format == "json":
            return self.to_json()
        raise ValueError(f"Unrecognized format '{self.format}'")


class ResultJson(Result):
    """Object that holds a json type result"""

    def __init__(self, json_result, **kwargs):
        self.json_result = json_result
        super().__init__(**kwargs)

    def to_pandas(self, index=None, sort=None):
        if isinstance(self.json_result, list):
            dataframe = DataFrame(self.json_result)
        else:
            dataframe = DataFrame([self.json_result])
        if sort:
            dataframe.sort_values(sort, inplace=True)
        if index:
            dataframe.set_index(index, inplace=True)
        return dataframe

    def to_votable(self):
        return Table(self.json_result)

    def to_json(self):
        return self.json_result

    def to_csv(self):
        df = self.to_pandas()
        return df.to_csv(index=False)


class ResultCsv(Result):
    """Object that holds a csv type result"""

    def __init__(self, csv_result_byte, **kwargs):
        self.csv_result = csv_result_byte.decode("utf-8")
        self.data = [x.split(",") for x in self.csv_result.split("\n")[1:-1]]

        self.columns = [x for x in self.csv_result.split("\n")[0].split(",")]
        super().__init__(**kwargs)

    def _rename_duplicates(self):
        counts = {}
        columns = self.columns.copy()
        for i, column in enumerate(columns):
            if column in counts:
                counts[column] += 1
                columns[i] = f"{column}_{counts[column] - 1}"
            else:
                counts[column] = 1
        return columns

    def to_pandas(self, index=None, sort=None):
        dataframe = read_csv(StringIO(self.csv_result))
        if sort:
            dataframe.sort_values(sort, inplace=True)
        if index:
            dataframe.set_index(index, inplace=True)
        return dataframe

    def to_votable(self):
        # TODO: Check if renaming the columns doesn't cause problems to the user
        columns = self._rename_duplicates()
        df = read_csv(StringIO(self.csv_result), names=columns, skiprows=1)
        df = df.convert_dtypes()
        table = Table.from_pandas(df)
        return table

    def to_json(self):
        columns = self._rename_duplicates()
        df = read_csv(StringIO(self.csv_result), names=columns, skiprows=1)
        return df.to_json(orient="records")

    def to_csv(self):
        return self.csv_result


class Client:
    def __init__(self, **kwargs):
        self.session = requests.Session()
        self.config = {}
        self.config.update(kwargs)
        self.allowed_formats = ["pandas", "votable", "json", "csv"]

    def load_config_from_file(self, path):
        pass

    def load_config_from_object(self, object):
        self.config.update(object)

    def _validate_format(self, format):
        format = format.lower()
        if format not in self.allowed_formats:
            raise FormatValidationError(
                "Format '%s' not in %s" % (format, self.allowed_formats), code=500
            )
        return format

    def _request(
        self,
        method,
        url,
        params=None,
        data=None,
        response_field=None,
        result_format="json",
        response_format="json",
    ):
        result_format = self._validate_format(result_format)

        resp = self.session.request(method, url, params=params, data=data)
        if resp.status_code >= 400:
            handle_error(resp, response_format)

        if response_format == "csv":
            return ResultCsv(resp.content, format=result_format)
        if response_field and result_format != "json" and result_format != "csv":
            return ResultJson(resp.json()[response_field], format=result_format)
        return ResultJson(resp.json(), format=result_format)
