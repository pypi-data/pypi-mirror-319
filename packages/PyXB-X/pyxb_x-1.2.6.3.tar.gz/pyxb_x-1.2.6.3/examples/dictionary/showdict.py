# -*- coding: utf-8 -*-

from __future__ import print_function
import sys
import dict
from pyxb.utils.six.moves.urllib.request import urlopen
import pyxb.utils.domutils as domutils
from xml.dom import minidom

# Get the list of dictionaries available from the service.
port_uri = 'http://services.aonaware.com/DictService/DictService.asmx'
uri = port_uri + '/DictionaryList'
dle_xml = urlopen(uri).read()
dle_dom = domutils.StringToDOM(dle_xml)
dle = dict.ArrayOfDictionary.createFromDOM(dle_dom)

op_path = '/DictionaryInfo'
for d in dle.Dictionary:
    # Create a REST-style query to retrieve the information about this dictionary.
    uri = '%s%s?dictId=%s' % (port_uri, op_path, d.Id)
    resp = urlopen(uri).read()
    print("%s (%s) : %d chars" % (d.Name, d.Id, len(resp)));

    # The response is a simple type derived from string, so we can
    # just extract and print it.  Excluded by default since it has
    # leading and trailing whitespace that causes problems with using
    # git to store the expected output.
    di_resp = dict.CreateFromDOM(domutils.StringToDOM(resp))
    if sys.stdout.isatty():
        print("%s\n" % di_resp);
