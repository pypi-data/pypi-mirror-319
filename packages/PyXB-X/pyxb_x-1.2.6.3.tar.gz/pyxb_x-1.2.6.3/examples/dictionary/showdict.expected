THE DEVIL'S DICTIONARY ((C)1911 Released April 15 1993) (devils) : 3235 chars
Easton's 1897 Bible Dictionary (easton) : 2182 chars
Elements database 20001107 (elements) : 1285 chars
The Free On-line Dictionary of Computing (27 SEP 03) (foldoc) : 3952 chars
U.S. Gazetteer (1990) (gazetteer) : 604 chars
The Collaborative International Dictionary of English v.0.44 (gcide) : 3168 chars
Hitchcock's Bible Names Dictionary (late 1800's) (hitchcock) : 1021 chars
Jargon File (4.3.1, 29 Jun 2001) (jargon) : 2992 chars
Virtual Entity of Relevant Acronyms (Version 1.9, June 2002) (vera) : 1355 chars
WordNet (r) 2.0 (wn) : 2562 chars
CIA World Factbook 2002 (world02) : 2544 chars
